### PR TITLE
Fix checking for too many function call arguments.

### PR DESCRIPTION
**Pull request by tritao**
This PR adds explicit checking for the case of too many function call arguments in function application instantiation.

This needs to be done before type checking because after type checking we drop the extra function call arguments due to having no type to check against.

During this PR development, it was found that Sway was allowing associated functions to be called with method call syntax,
which should not be allowed (Rust also throws an error in this case), and previously was just working by chance and a side
effect of the zip calls dropping the extra arguments.

So this adds a new error and check for this and updates some tests that were relying on this non-Rust-like behaviour.

There was another issue found, where the IR (at least the IR parser) currently cannot deal with empty structs, which was triggered by an existing IR test after being updated to work with the above changes.

Also re-enables the `should_fail/excess_fn_arguments` test.

Closes https://github.com/FuelLabs/sway/issues/2081.

### CODE REVIEW

**Comment by vaivaswatha** [comment]
>This needs to be done before type checking because after type checking we drop the extra function call arguments due to having no type to check against.

Does this mean that, with your changes, `instantiate_function_application_inner` can never see `TooManyArgumentsForFunction`? That means we could also check for `TooFewArgumentsForFunction` in the outer functions and remove the checks entirely from `instantiate_function_application_inner` ?

File: sway-core/tests/sway_to_ir/get_storage_key.sw [hunk 1]
```diff
@@ -1,16 +1,18 @@
 contract;
 
-struct Empty {}
+struct Foo {
+    foo: u32
```
Comment (mohammadfawaz) [inline sway-core/tests/sway_to_ir/get_storage_key.sw:4]: As discussed, please create an issue to investigate the need for this change and the issues with the IR parser.

**Review by mohammadfawaz — CHANGES_REQUESTED** [review]

**Review by vaivaswatha — DISMISSED** [review]

**Comment by tritao** [comment]
Updated the PR with fixes to the IR parsing issue instead of changing the test.

Meanwhile I've also found a couple more issues with associated methods while reworking the checks I had earlier on this PR.

The new checking code is more robust and found a couple more issues related to traits, as such I've decided to just split it from this one, and will be sending a different PR later to deal with that with more tests once I handle those extra edge cases.

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/function_application.rs [hunk 2]
```diff
@@ -107,59 +120,71 @@ pub(crate) fn instantiate_function_application_simple(
...
+    selector: Option<ContractCallMetadata>,
+    is_constant: IsConstant,
+    self_state_idx: Option<StateIndex>,
+    span: Span,
+) -> TypedExpression {
+    assert_eq!(
+        arguments.len(),
+        function_decl.parameters.len(),
+        "different number of function parameters and arguments"
+    );
```
Comment (otrho) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression/function_application.rs:173]: We tend not to have assertions in the compiler as permanent code (sometimes there'll be a `todo!()` or `unimplemented!()` temporarily) but instead prefer to return a `CompileError::Internal` to encourage users to file an issue if something like this occurs.

File: sway-ir/src/parser.rs [hunk 3]
```diff
@@ -389,8 +389,8 @@ mod ir_builder {
                 }
 
             rule struct_const() -> IrAstConstValue
-                = "{" _ flds:(field_or_element_const() ++ comma()) "}" _ {
-                    IrAstConstValue::Struct(flds)
+                = "{" _ flds:(field_or_element_const() ++ comma())? "}" _ {
+                    IrAstConstValue::Struct(flds.unwrap_or_default())
```
Comment (otrho) [inline sway-ir/src/parser.rs:393]: A better way to do this is with the `**` operator.  It's a bit more explicit and doesn't need the `unwrap()`.

**Review by otrho — COMMENTED** [review]
Looking good, just a couple of changes needed.

File: sway-core/tests/sway_to_ir/get_storage_key.sw (hunk 1 above)
Comment (otrho) [reply sway-core/tests/sway_to_ir/get_storage_key.sw]: Yep, I don't think this has been done yet?

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/method_application.rs [hunk 4]
```diff
@@ -242,6 +245,14 @@ pub(crate) fn type_check_method_application(
                 None
             };
 
+            if method.parameters.is_empty() && !method.is_contract_call {
```
[the remark this answers is not in the document]
Comment (mohammadfawaz) [reply sway-core/src/semantic_analysis/ast_node/expression/typed_expression/method_application.rs]: I guess this is no longer relevant.

**Review by mohammadfawaz — APPROVED** [review]
LGTM! Thanks.

File: sway-core/tests/sway_to_ir/get_storage_key.sw (hunk 1 above)
Comment (mohammadfawaz) [reply sway-core/tests/sway_to_ir/get_storage_key.sw]: I guess this particular change is no longer there but an issue about empty structs in the IR and how they work if they're passed around etc. would still be useful.

**Review by otrho — APPROVED** [review]

**Pull request merged**